design/kata-api-design.md

```diff
@@ -0,0 +1,112 @@
+|sandbox.UpdateNetwork()| Update the configuration of an existing network interface.|
+|sandbox.UpdateRoute()| Update the sandbox route table, e.g. for portmapping support.|
+
+### Sandbox Relay API
+|Name|Description|
+|---|---|
+|sandbox.WinsizeProcess(containerID, processID, Height, Width)|Relay TTY resize request to a process.|
+|sandbox.SignalProcess(containerID, processID, signalID, signalALL)| Relay a signal to a process or all processes in a container.|
+|sandbox.IOStream(containerID, processID)| Relay a process stdio. Return stdin/stdout/stderr pipes to the process stdin/stdout/stderr streams.|
+
+## Plugin framework for external proprietary Kata runtime extensions
+### Hypervisor plugin
+
+TBD.
+### Metadata storage plugin
+The metadata storage plugin controls where sandbox metadata is saved.
+All metadata storage plugins must implement following API:
+
+|Name|Description|
+|---|---|
+|storage.Save(key, value)| Save a record.|
+|storage.Load(key)| Load a record.|
+|storage.Delete(key)| Delete a record.|
+
+Built-in implementations include:
+   -  Filesystem storage
+   -  LevelDB storage
+
+### VM Factory plugin
+The VM factory plugin controls how a sandbox factory creates new VMs.
+All VM factory plugins must implement following API:
+
+|Name|Description|
+|---|---|
+|VMFactory.NewVM(HypervisorConfig)|Create a new VM based on `HypervisorConfig`.|
+
+Built-in implementations include:
+
+|Name|Description|
+|---|---|
+|CreateNew()| Create brand new VM based on HypervisorConfig|
+|CreateFromTemplate()| Create new VM from template.|
+|CreateFromCache()| Create new VM from VM caches.|
+
+### Sandbox Creation Plugin Workflow
+![Sandbox Creation Plugin Workflow](https://raw.githubusercontent.com/bergwolf/raw-contents/master/kata/Kata-sandbox-creation.png "Sandbox Creation Plugin Workflow")
+
+### Sandbox Connection Plugin Workflow
+![Sandbox Connection Plugin Workflow](https://raw.githubusercontent.com/bergwolf/raw-contents/master/kata/Sandbox-Connection.png "Sandbox Connection Plugin Workflow")
+
+## Built-in Shim and Proxy Types and Capabilities
+### Built-in shim/proxy sandbox configurations
+-  Supported Shim configurations:
+
+|Name|Description|
+|---|---|
+|noopshim|do not start any shim process.|
+|ccshim| start the cc-shim binary.|
+|katashim| start the kata-shim binary.|
+-  Supported Proxy configurations:
+
+|Name|Description|
+|---|---|
+|noopProxy| a dummy proxy implementation of the proxy interface, only used for testing purpose.|
+|noProxy|generic implementation for any case where no actual proxy is needed.|
+|ccProxy|run ccProxy to proxy between runtime and agent.|
+|kataProxy|run kata-proxy to translate yamux connections between runtime and kata agent. |
+|kataNoProxy| no standalone proxy process and connect to kata agent with internal yamux translation.|
+
+### Built-in Shim Capability
+Built-in shim capability is implemented by removing standalone shim process, and
+supporting the shim related APIs.
```

Comment:
Please add this would be a `kata_shim_builtin` implementation, given the fact that we already have `noop_shim`, `cc-shim` and `kata_shim`.

Reply:
Good point! Thanks!

Reply:
Good point! Thanks!